File: 计算签名/HttpDemo.java
/**
 * @author erniu.wzh
 * @date 2021/1/8 2:06 下午
 */

import com.alibaba.fastjson.JSONObject;
import org.apache.commons.codec.binary.Base64;
import org.apache.commons.codec.digest.DigestUtils;
import org.apache.http.Header;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public class HttpDemo {
    public static void main(String[] args) throws InvalidKeyException, NoSuchAlgorithmException {
        String umAppkey = "xxxx";
        String appKey = "xxxx";
        String appSecret = "xxxx";
        String token = "xxxx";
	// 下面的url要和阿里云云市场购买的商品对应
        String url = "https://verify5.market.alicloudapi.com/api/v1/mobile/info?appkey=" + umAppkey;
        HttpPost httpPost = new HttpPost(url);
        /**
         * body
         */
        JSONObject object = new JSONObject();
        object.put("token", token);
        StringEntity stringEntity = new StringEntity(object.toJSONString(), StandardCharsets.UTF_8);
        httpPost.setEntity(stringEntity);
        /**
         * header
         */
        httpPost.setHeader("Content-Type", "application/json; charset=UTF-8");
        httpPost.setHeader("Accept", "application/json");
        httpPost.setHeader("X-Ca-Version", "1");
        httpPost.setHeader("X-Ca-Signature-Headers", "X-Ca-Version,X-Ca-Stage,X-Ca-Key,X-Ca-Timestamp");
        httpPost.setHeader("X-Ca-Stage", "RELEASE");
        httpPost.setHeader("X-Ca-Key", appKey);
        httpPost.setHeader("X-Ca-Timestamp", String.valueOf(System.currentTimeMillis()));
        httpPost.setHeader("X-Ca-Nonce", UUID.randomUUID().toString());
        httpPost.setHeader("Content-MD5", Base64.encodeBase64String(DigestUtils.md5(object.toJSONString())));
        /**
         * sign
         */
        String stringToSign = getSignString(httpPost);
        Mac hmacSha256 = Mac.getInstance("HmacSHA256");
        byte[] keyBytes = appSecret.getBytes(StandardCharsets.UTF_8);
        hmacSha256.init(new SecretKeySpec(keyBytes, 0, keyBytes.length, "HmacSHA256"));
        String sign = new String(Base64.encodeBase64(hmacSha256.doFinal(stringToSign.getBytes(StandardCharsets.UTF_8))));
        httpPost.setHeader("X-Ca-Signature", sign);
        /**
         * execute
         */
        CloseableHttpClient httpclient = HttpClients.createDefault();
        try {
            CloseableHttpResponse response = httpclient.execute(httpPost);
            System.out.println(response);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static String getSignString(HttpPost httpPost) {
        Header[] headers = httpPost.getAllHeaders();
        Map<String, String> map = new HashMap<>();
        for (Header header : headers) {
            map.put(header.getName(), header.getValue());
        }
        return httpPost.getMethod() + "\n" +
                map.get("Accept") + "\n" +
                map.get("Content-MD5") + "\n" +
                map.get("Content-Type") + "\n\n" +
                "X-Ca-Key:" + map.get("X-Ca-Key") + "\n" +
                "X-Ca-Stage:" + map.get("X-Ca-Stage") + "\n" +
                "X-Ca-Timestamp:" + map.get("X-Ca-Timestamp") + "\n" +
                "X-Ca-Version:" + map.get("X-Ca-Version") + "\n" +
                httpPost.getURI().getPath() + "?" + httpPost.getURI().getQuery();
    }
}

File: 计算签名/HttpDemo.php
<?php
function oneclick_login()
{
    $mobile_token = $_POST['mobile_token'];
    $host = "https://verify5.market.alicloudapi.com";
    $path = "/api/v1/mobile/info";
    $method = "POST";
    $accept = "application/json";
    $content_type = "application/json; charset=UTF-8";
    $appKey = ""; //阿里云市场购买应用的appKey
    $appSecret = ""; //阿里云市场购买应用的appSecret
    $um_appkey = ""; //友盟的appkey

    $header["Accept"] = $accept;
    $header["Content-Type"] = $content_type;
    $header["X-Ca-Version"] = "1";
    $header["X-Ca-Signature-Headers"] = "X-Ca-Key,X-Ca-Nonce,X-Ca-Stage,X-Ca-Timestamp,X-Ca-Version";
    $header["X-Ca-Stage"] = "RELEASE";
    $header["X-Ca-Key"] = $appKey; //请求的阿里云AppKey
    $header["X-Ca-Timestamp"] = strval(time() * 1000);
    mt_srand((double) microtime() * 10000);
    $uuid = strtoupper(md5(uniqid(rand(), true)));
    $header["X-Ca-Nonce"] = strval($uuid);
    //Headers
    $headers = "X-Ca-Key:" . $header["X-Ca-Key"] . "\n";
    $headers .= "X-Ca-Nonce:" . $header["X-Ca-Nonce"] . "\n";
    $headers .= "X-Ca-Stage:" . $header["X-Ca-Stage"] . "\n";
    $headers .= "X-Ca-Timestamp:" . $header["X-Ca-Timestamp"] . "\n";
    $headers .= "X-Ca-Version:" . $header["X-Ca-Version"] . "\n";
    //Url
    $url = $path . "?appkey=" . $um_appkey; //appkey为用户在友盟注册的应用分配的appKey,token和phoneNumber是app传过来的值
    //sign
    $str_sign = $method . "\n";
    $str_sign .= $accept . "\n";
    $str_sign .= "\n";
    $str_sign .= $content_type . "\n";
    $str_sign .= "\n";
    $str_sign .= $headers;
    $str_sign .= $url;
    $sign = base64_encode(hash_hmac('sha256', $str_sign, $appSecret, true)); //secret为APP的密钥
    $header['X-Ca-Signature'] = $sign;
    $post_data['token'] = $mobile_token;
    //curl
    $curl_url = $host . $path . "?appkey=" . $um_appkey; //appkey为用户在友盟注册的应用分配的appKey
    $headerArray = array();
    foreach ($header as $k => $v) {
        array_push($headerArray, $k . ":" . $v);
    }
    $curl = curl_init();
    curl_setopt($curl, CURLOPT_CUSTOMREQUEST, $method);
    curl_setopt($curl, CURLOPT_URL, $curl_url);
    curl_setopt($curl, CURLOPT_HTTPHEADER, $headerArray);
    curl_setopt($curl, CURLOPT_FAILONERROR, false);
    curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($curl, CURLOPT_TIMEOUT, 10);
    curl_setopt($curl, CURLOPT_CONNECTTIMEOUT, 5);
    curl_setopt($curl, CURLOPT_HEADER, true);
    curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, false);
    curl_setopt($curl, CURLOPT_SSL_VERIFYHOST, false);
    curl_setopt($curl, CURLOPT_POST, 1);
    curl_setopt($curl, CURLOPT_POSTFIELDS, json_encode($post_data));
    $result = curl_exec($curl);
}
